k8s hook runner fails occasionally

Hello,
We're having an issue on our self hosted runner, we're using k8s hook to execute our jobs, but some of jobs fails occasionally  with this error, can't seems to understand what the issue is.

```
##[debug]/runner/externals/node16/bin/node /runner/k8s/index.js
node:internal/process/promises:246
          triggerUncaughtException(err, true /* fromPromise */);
          ^
[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "#<ErrorEvent>".] {
  code: 'ERR_UNHANDLED_REJECTION'
}
Error: Process completed with exit code 1.
Error: Executing the custom container implementation failed. Please contact your self hosted runner administrator.
```
Image: summerwind/actions-runner:v2.296.1-ubuntu-20.04
GHES version: 3.6.1
Any idea what the issue is ?
Thank you

Hi @nikola-jokic,
I can give more information about this issue _(I'm working with @atahardjebbar-ledger)_, so we are still experiencing this issue for almost all workflow runs (with 0 to 5 jobs failing), and we build+run the version with the fix (#49).

Regarding the workflow file, I can't share it publicly, but It's composed of multiple jobs/actions to do the compilation. This workflow is composed of around 20 jobs, each job having a runtime from 5 min to 1h30 (with a total time for the workflow of approximately 4h).

So, we have more information about the error based on the debug log (sent to Github); in this case, it happened for the "Post-Clone" operation:
```
##[debug]Evaluating condition for step: 'Post Clone'
##[debug]Evaluating: always()
##[debug]Evaluating always:
##[debug]=> true
##[debug]Result: true
##[debug]Starting: Post Clone
##[debug]Loading inputs
##[debug]Evaluating: github.ref
##[debug]Evaluating Index:
##[debug]..Evaluating github:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'ref'
##[debug]=> 'refs/heads/next'
##[debug]Result: 'refs/heads/next'
##[debug]Evaluating: github.repository
##[debug]Evaluating Index:
##[debug]..Evaluating github:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'repository'
##[debug]=> '<REDACTED>'
##[debug]Result: '<REDACTED>'
##[debug]Evaluating: github.token
##[debug]Evaluating Index:
##[debug]..Evaluating github:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'token'
##[debug]=> '***'
##[debug]Result: '***'
##[debug]Loading env
Post job cleanup.
##[debug]Running JavaScript Action with default external tool: node1[2](https://<REDACTED>/runs/62041?check_suite_focus=true#step:24:2)
Run '/runner/k8s/index.js'
  shell: /runner/externals/node16/bin/node {0}
##[debug]/runner/externals/node16/bin/node /runner/k8s/index.js
node:internal/process/promises:2[4](https://<REDACTED>/runs/62041?check_suite_focus=true#step:24:4)[6](https://<REDACTED>/runs/62041?check_suite_focus=true#step:24:6)
          triggerUncaughtException(err, true /* fromPromise */);
          ^

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "#<ErrorEvent>".] {
  code: 'ERR_UNHANDLED_REJECTION'
}
Error: Process completed with exit code 1.
Error: Executing the custom container implementation failed. Please contact your self hosted runner administrator.
##[debug]System.Exception: Executing the custom container implementation failed. Please contact your self hosted runner administrator.
##[debug] ---> System.Exception: The hook script at '/runner/k[8](https://<REDACTED>/runs/62041?check_suite_focus=true#step:24:8)s/index.js' running command 'RunScriptStep' did not execute successfully
##[debug]   at GitHub.Runner.Worker.Container.ContainerHooks.ContainerHookManager.ExecuteHookScript[T](IExecutionContext context, HookInput input, ActionRunStage stage, String prependPath)
##[debug]   --- End of inner exception stack trace ---
##[debug]   at GitHub.Runner.Worker.Container.ContainerHooks.ContainerHookManager.ExecuteHookScript[T](IExecutionContext context, HookInput input, ActionRunStage stage, String prependPath)
##[debug]   at GitHub.Runner.Worker.Container.ContainerHooks.ContainerHookManager.RunScriptStepAsync(IExecutionContext context, ContainerInfo container, String workingDirectory, String entryPoint, String entryPointArgs, IDictionary`2 environmentVariables, String prependPath)
##[debug]   at GitHub.Runner.Worker.Handlers.ContainerStepHost.ExecuteAsync(IExecutionContext context, String workingDirectory, String fileName, String arguments, IDictionary`2 environment, Boolean requireExitCodeZero, Encoding outputEncoding, Boolean killProcessOnCancel, Boolean inheritConsoleHandler, String standardInInput, CancellationToken cancellationToken)
##[debug]   at GitHub.Runner.Worker.Handlers.NodeScriptActionHandler.RunAsync(ActionRunStage stage)
##[debug]   at GitHub.Runner.Worker.ActionRunner.RunAsync()
##[debug]   at GitHub.Runner.Worker.StepsRunner.RunStepAsync(IStep step, CancellationToken jobCancellationToken)
##[debug]Finishing: Post Clone
```

Thanks for your help; also, do you have a recommended way to extract logs from the `_diag` folder? (because we are using ephemeral runner, so each time a job finishes, logs are gone)

--- a/packages/k8s/src/index.ts
+++ b/packages/k8s/src/index.ts
@@ -38,7 +38,6 @@ async function run(): Promise<void> {
       case Command.RunContainerStep:
         exitCode = await runContainerStep(args)
         return process.exit(exitCode)
-      case Command.runContainerStep:
       default:
         throw new Error(`Command not recognized: ${command}`)
     }

--- a/packages/k8s/src/index.ts
+++ b/packages/k8s/src/index.ts
@@ -9,7 +9,6 @@ import {
 import { isAuthPermissionsOK, namespace, requiredPermissions } from './k8s'
 
 async function run(): Promise<void> {
-  let exitCode = 0
   try {
     const input = await getInputFromStdin()
 
@@ -24,28 +23,29 @@ async function run(): Promise<void> {
         )} on the pod resource in the '${namespace()}' namespace. Please contact your self hosted runner administrator.`
       )
     }
+
+    let exitCode = 0
     switch (command) {
       case Command.PrepareJob:
         await prepareJob(args as prepareJobArgs, responseFile)
-        break
+        return process.exit(0)
       case Command.CleanupJob:
         await cleanupJob()
-        break
+        return process.exit(0)
       case Command.RunScriptStep:
         await runScriptStep(args, state, null)
-        break
+        return process.exit(0)
       case Command.RunContainerStep:
         exitCode = await runContainerStep(args)
-        break
+        return process.exit(exitCode)
       case Command.runContainerStep:
       default:
         throw new Error(`Command not recognized: ${command}`)
     }
   } catch (error) {
     core.error(error as Error)
-    exitCode = 1
+    process.exit(1)
   }
-  process.exitCode = exitCode
 }
 
 void run()

